language/diem-framework/modules/DiemAccount.move

```diff
@@ -774,27 +778,39 @@ module DiemAccount {
         ensures spec_holds_own_withdraw_cap(cap_addr);
     }
 
+    public fun pay_from<Token: store>(
+        cap: &WithdrawCapability,
+        payee: address,
+        amount: u64,
+        metadata: vector<u8>,
+        metadata_signature: vector<u8>,
+    ) acquires DiemAccount, Balance, AccountOperationsCapability {
+        pay<Token>(cap, payee, amount, metadata, metadata_signature, true);
+    }
+
     /// Withdraw `amount` Diem<Token> from the address embedded in `WithdrawCapability` and
     /// deposits it into the `payee`'s account balance.
     /// The included `metadata` will appear in the `SentPaymentEvent` and `ReceivedPaymentEvent`.
     /// The `metadata_signature` will only be checked if this payment is subject to the dual
-    /// attestation protocol
-    public fun pay_from<Token: store>(
+    /// attestation protocol and `dual_attestation` is true.
+    public fun pay<Token: store>(
         cap: &WithdrawCapability,
         payee: address,
         amount: u64,
         metadata: vector<u8>,
-        metadata_signature: vector<u8>
+        metadata_signature: vector<u8>,
+        dual_attestation: bool
```

Comment:
Although this is ok in today's Diem, this would not be safe in a system with open publishing--devs can opt of dual attestation by passing `false` here. Here's a half-baked idea for an alternative strategy without the bool:


```
// in DualAttestation.move

/// Created iff payee a transfer of `token` from `payer to `payee` has been approved
struct Approved<Token: store> { payer: address, payee: address, token: Diem<Token> }

// For the old dual attestation code
public fun legacy_check<Token: store>(
  payer: address, payee: address, token: Diem<Token>, metadata: vector<u8>, metadata_signature_vector<u8>
): Approved {
    assert_payment_ok<Token>(payer, payee, Diem::value(&token), metadata, metadata_signature);
    Approved { payer, payee, token }
}

// for the new multi-agent dual attestation code
public fun check<Token: store>(payer: &signer, payee: &signer, token: Diem<Token>): DualAttestationOk {
    Approved { payer: Signer::address_of(payer), payee: Signer::address_of(payee), token }
}

pub fun into<Token: store>(approved: Approved<Token): (address, address, Diem<Token>) {
  let Approver { payer, payee, token } = approved;
  (payer, payee, token)
}
```

Then, you have `DiemAccount::deposit` take a `DualAttestation::Approved` resource as input and can unpack it to get the coin to deposit + the payer/payee addresses to use in events.

Reply:
Isn't someone can do the same? by passing in the Approved to new deposit function to bypass dual attestation?

Reply:
Further more, you can also direct call `Diem::deposit` to bypass dual attestation.

Reply:
Ah, good question. They cannot do that because the only way to get an `Approved` resource is by calling `check` or `legacy_check` (which will perform the dual attestation checks).

Reply:
For bypassing the dual attestation, you can first split the amount into smaller ones, then call pay_from to transfer.

Reply:
Yes, fair point (this strategy would work both before and after this PR). The only thing the TR is intended to enforce is that each individual `deposit` complies with the TR rules.